File: python/euler/bin/p0096.py
# project euler: problem 96

#   [Grid]
#       C0 C1 C2 C3 C4 C5 C6 C7 C8
#      +--------+--------+--------+    R: Row
#    R0|        |        |        |    C: Column
#    R1|   B0   |   B1   |   B2   |    B: Box
#    R2|        |        |        |
#      +--------+--------+--------+    Cell = RxCy
#    R3|        |        |        |
#    R4|   B3   |   B4   |   B5   |
#    R5|        |        |        |
#      +--------+--------+--------+
#    R6|        |        |        |
#    R7|   B6   |   B7   |   B8   |
#    R8|        |        |        |
#      +--------+--------+--------+

import re
from functools import reduce
from typing import IO

from euler.lib.util import flatten

# row/column/position(cell)
ROW = ['R' + str(i) for i in range(9)]  # R0, R1, ..., R8
COL = ['C' + str(i) for i in range(9)]  # C0, C1, ..., C8
POS = [r + c for r in ROW for c in COL]  # R0C0, R0C1, ..., R8C8

# list of cells belonging to row/column/box
ROW_LST = [[r + c for c in COL] for r in ROW]
COL_LST = [[r + c for r in ROW] for c in COL]
BOX_LST = [
    [r + c for r in rows for c in cols]
    for rows in (ROW[:3], ROW[3:6], ROW[6:])
    for cols in (COL[:3], COL[3:6], COL[6:])
]

# dictionary of cells and their corresponding row, column and box
groups = dict(
    (p, [group for group in ROW_LST + COL_LST + BOX_LST if p in group]) for p in POS
)
# dictionary of cells and their corresponding cells
links = dict((p, set(flatten(groups[p])) - set((p,))) for p in POS)


class Grid:
    def __init__(self, data: str):
        self._data = data

    # We make a grid as dictionary which values are immutable data, string.
    # The result is an easy backtracked search with a shallow copy of the dictionary.
    def setup_grid(self) -> bool:
        self._grid = dict((p, '123456789') for p in POS)
        for pos, ch in zip(POS, self._data):
            if ch != '0':
                if self._decide_num(self._grid, pos, ch) is None:
                    return False

        return True

    # Tentatively determine a number to leave in the cell and remove the other numbers
    def _decide_num(
        self, grid: dict[str, str], pos: str, num: str
    ) -> dict[str, str] | None:
        # Remove all numbers from the cell except the one we have decided on
        other_nums = grid[pos].replace(num, '')
        if all(self._remove_num(grid, pos, n) for n in other_nums):
            return grid
        else:
            return None

    # Remove the number from the candidate in the cell and update involved cells
    #
    # Note: After processing this method, the calling variable 'grid' is updated
    #       since this dictionary parameter is not copied from the original
    #       but is a reference to the original.
    def _remove_num(self, grid: dict[str, str], pos: str, num: str) -> bool:
        # If the target number to remove is not already in the cell, do nothing
        if num not in grid[pos]:
            return True
        # If there is only the target number in the cell to be removed, it is a contradiction
        if grid[pos] == num:
            return False

        grid[pos] = grid[pos].replace(num, '')
        # If the deletion results only one number is in the cell,
        # remove the number from the linked cells
        if len(grid[pos]) == 1:
            if (
                all(self._remove_num(grid, p, grid[pos]) for p in links[pos])
                is not True
            ):
                return False

        # check row/column/box
        for group in groups[pos]:
            # If the target number doesn't remain in a belong row/column/box,
            # it is a contradiction.
            cells = [sq for sq in group if num in grid[sq]]
            if not cells:
                return False
            # If there is an only one cell which contains the removed number
            # in a belong row/column/box, the number is decided tentatively
            # to remain in the cell
            if len(cells) == 1:
                if self._decide_num(grid, cells[0], num) is None:
                    return False

        # everything is OK
        return True

    def _solve(self, grid: dict[str, str] | None) -> dict[str, str] | None:
        if grid is None:
            return None

        cells = [(len(grid[p]), p) for p in POS]
        if all(c[0] == 1 for c in cells):
            return grid

        _, pos = min((c for c in cells if c[0] > 1), key=lambda x: x[0])
        for num in grid[pos]:
            if (
                result := self._solve(self._decide_num(grid.copy(), pos, num))
            ) is not None:
                return result

        return None

    def solve(self) -> dict[str, str] | None:
        return self._solve(self._grid)


def parse_data(fh: IO) -> list[str]:
    def trim(lst):
        s = reduce(lambda x, y: x + y, lst)
        tmp = re.sub(r'[^0-9.]', '', s)
        return re.sub(r'\.', '0', tmp)

    result = []
    acc = []
    for line in fh.read().splitlines():
        if re.match(r'[0-9.]', line) is not None:
            acc.append(line)
        elif re.match(r'-', line) is not None or not acc:
            continue
        else:
            result.append(trim(acc))
            acc = []
    else:
        if len(acc) == 9:
            result.append(trim(flatten(acc)))

    assert all([len(s) == 81 for s in result]), 'invalid data file'
    return result


def compute(fh: IO) -> str:
    puzzles = parse_data(fh)
    acc = 0
    for idx, problem in enumerate(puzzles):
        grid = Grid(problem)
        if not grid.setup_grid():
            assert False, 'invalid data: Grid {}'.format(idx + 1)

        d = grid.solve()
        if d is not None:
            acc += int(d['R0C0'] + d['R0C1'] + d['R0C2'])
        else:
            print('Warning: No answer found: Grod {}'.format(idx + 1))

    return str(acc)


def solve() -> str:
    from euler.lib.resource import asset_file

    fh = asset_file('p096_sudoku.txt')
    result = compute(fh)
    fh.close()
    return result

File: typescript/euler/bin/p0096.ts
// project euler: problem 96

/*
  [Grid]
      C0 C1 C2 C3 C4 C5 C6 C7 C8
     +--------+--------+--------+    R: Row
   R0|        |        |        |    C: Column
   R1|   B0   |   B1   |   B2   |    B: Box
   R2|        |        |        |
     +--------+--------+--------+    Cell = RxCy
   R3|        |        |        |
   R4|   B3   |   B4   |   B5   |
   R5|        |        |        |
     +--------+--------+--------+
   R6|        |        |        |
   R7|   B6   |   B7   |   B8   |
   R8|        |        |        |
     +--------+--------+--------+

  We will need the following files to run this program.
    - https://projecteuler.net/project/resources/p096_sudoku.txt
*/

import { chunk, zip } from "@std/collections";
import { cartesianProduct } from "combinatorics/mod.ts";
import { assetData } from "../lib/asset.ts";
import { maxLst } from "../lib/math.ts";
import { dedupSort, range } from "../lib/util.ts";

/* row/column/position(cell) */

// R0, R1, ..., R8
const ROW = range(0, 9).map((x) => "R" + x);

// C0, C1, ..., C8
const COL = range(0, 9).map((x) => "C" + x);

// R0C0, R0C1, ..., R8C8
const POS = [...cartesianProduct(ROW, COL)].map(([r, c]) => r + c);

/* list of cells belonging to row/column/box */

const ROW_LST = ROW.map((r) => POS.filter((pos) => pos.includes(r)));
const COL_LST = COL.map((c) => POS.filter((pos) => pos.includes(c)));
const BOX_LST = [...cartesianProduct(chunk(ROW, 3), chunk(COL, 3))].map((
  [rs, cs],
) => [...cartesianProduct(rs, cs)]).map((lst) => lst.map(([r, c]) => r + c));

/* map of cells and their corresponding row, column and box */
const groups = new Map(
  POS.map((pos) => [
    pos,
    ROW_LST.concat(COL_LST).concat(BOX_LST).filter((lst) => lst.includes(pos)),
  ]),
);

/* map of cells and their corresponding cells */
const links = new Map(
  [...groups.entries()].map(([k, v]) => [
    k,
    dedupSort(v.flat().filter((x) => x !== k)),
  ]),
);

class Grid {
  // We make a grid as Map object which values are primitive data, string.
  // It make backtracking search easily.
  grid = new Map(POS.map((pos) => [pos, "123456789"]));
  data: string;

  constructor(data: string) {
    this.data = data;
  }

  setupGrid(): boolean {
    for (const [pos, ch] of zip(POS, this.data.split(""))) {
      if (ch !== "0") {
        if (!this.decideNum(this.grid, pos, ch)) {
          return false;
        }
      }
    }

    return true;
  }

  // Tentatively determine a number to leave in the cell and remove other numbers
  decideNum(
    grid: Map<string, string>,
    pos: string,
    num: string,
  ): Map<string, string> | undefined {
    const otherNums = grid.get(pos)!.replace(num, "");
    if (otherNums.split("").every((n) => this.removeNum(grid, pos, n))) {
      return grid;
    } else {
      return undefined;
    }
  }

  // Remove the number from the candidate in the cell and update involved cells
  //
  // Note: After processing this method, the calling variable 'grid' is updated
  //       since this parameter is not copied from the original
  //       but a reference to the original.
  removeNum(grid: Map<string, string>, pos: string, num: string): boolean {
    // If the target number to remove is not already in the cell, do nothing
    if (!grid.get(pos)!.includes(num)) {
      return true;
    }
    // If there is only the target number in the cell to be removed, it is a contradiction
    if (grid.get(pos) === num) {
      return false;
    }

    grid.set(pos, grid.get(pos)!.replace(num, ""));

    // If the deletion results only one number is in the cell,
    // remove the number from the linked cells
    if (grid.get(pos)!.length === 1) {
      if (
        !links.get(pos)?.every((p) => this.removeNum(grid, p, grid.get(pos)!))
      ) {
        return false;
      }
    }

    // check row/column/box
    for (const group of groups.get(pos)!) {
      const cells = group.filter((sq) => grid.get(sq)!.includes(num));

      // If the target number doesn't remain in a belong row/column/box, it is a contradiction.
      if (cells.length === 0) {
        return false;
      }

      // If there is an only one cell which contains the removed number in a belong row/column/box,
      // the number is decided tentatively to remain in the cell
      if (cells.length === 1) {
        if (!this.decideNum(grid, cells[0], num)) {
          return false;
        }
      }
    }

    // everything is OK.
    return true;
  }

  _solve(
    grid: Map<string, string> | undefined,
  ): Map<string, string> | undefined {
    if (grid === undefined) {
      return undefined;
    }

    // an answer found
    const cells: [number, string][] = POS.map((p) => [grid.get(p)!.length, p]);
    if (cells.every((c) => c[0] === 1)) {
      return grid;
    }

    // select the cell with the smallest number of candidates and continue the search
    const [_, pos] =
      cells.filter((c) => c[0] > 1).sort((a, b) => a[0] - b[0])[0];
    for (const num of grid.get(pos)!) {
      const result = this._solve(this.decideNum(new Map(grid), pos, num));
      if (result) {
        return result;
      }
    }

    return undefined;
  }

  solve(): Map<string, string> | undefined {
    return this._solve(this.grid);
  }

  // for debug
  display(grid: Map<string, string>) {
    const values = [...grid.values()];
    const width = maxLst(values.map((s) => s.length)) + 1;
    const line_sep = [
      "-".repeat(width * 3),
      "-".repeat(width * 3),
      "-".repeat(width * 3),
    ].join("+");

    let sep_flag = 0;
    for (const row of chunk(values.map((x) => x.padStart(width, " ")), 9)) {
      console.log(chunk(row, 3).map((x) => x.join("")).join("|"));
      sep_flag += 1;
      if (sep_flag === 3 || sep_flag === 6) {
        console.log(line_sep);
      }
    }
    console.log("");

    return;
  }
}

const parseData = (data: string): string[] => {
  const splitLines = (str: string): string[] => {
    const result = str.split(/\r?\n/);
    if (result.at(-1) === "") {
      return result.slice(0, -1);
    } else {
      return result;
    }
  };

  const trim = (lst: string[]): string =>
    lst.reduce((acc, cur) => acc + cur)
      .replace(/[^0-9.]/g, "")
      .replaceAll(".", "0");

  let acc: string[] = [];
  const result: string[] = [];

  for (const line of splitLines(data)) {
    if (line.match(/^[0-9.]/) !== null) {
      acc.push(line);
    } else if (line.match(/^-/) !== null || acc.length === 0) {
      continue;
    } else {
      result.push(trim(acc));
      acc = [];
    }
  }
  if (acc.length === 9) {
    result.push(trim(acc));
  }

  if (result.some((x) => x.length !== 81)) {
    throw new Error("invalid data file");
  }

  return result;
};

export const compute = (data: string): string => {
  let acc = 0;
  for (const [idx, problem] of parseData(data).entries()) {
    const grid = new Grid(problem);
    if (!grid.setupGrid()) {
      console.log("Invalid data: Grid", idx + 1);
      continue;
    }

    const d = grid.solve();
    if (d) {
      acc += Number(d.get("R0C0")! + d.get("R0C1")! + d.get("R0C2")!);
    } else {
      console.log("No answer found: Grod", idx + 1);
    }
  }

  return String(acc);
};

export const solve = (): string => {
  const data = new TextDecoder().decode(assetData("p096_sudoku.txt"));
  return compute(data);
};
